File: Auxiliar 1 - 2019P/python/cc3002/auxiliares2019p/aux1/unit.py
"""
This module contains the implementation of a unit from a generic game.
"""


def attack(enemy: 'Unit', damage: int):
    enemy.hit_points -= damage


class Unit:
    """
    Base class for the units
    """
    _hit_points: int
    _strength: int

    def __init__(self, hit_points: int, strength: int):
        self._hit_points = hit_points
        self._strength = strength

    @property
    def hit_points(self):
        return self._hit_points

    @hit_points.setter
    def hit_points(self, value):
        self._hit_points = value

    def attack(self, enemy: 'Unit'):
        pass

File: Auxiliar 1 - 2019P/kotlin/cc3002/auxiliares2019p/aux1/Unit.kt
package cc3002.auxiliares2019p.aux1

/**
 * Common interface for all units.
 *
 * @author Ignacio Slater Muñoz
 * @since 0.0.x01
 */
interface IUnit {
    var hitPoints: Int
    val strength: Int

    /**
     * Attacks another unit.
     *
     * @param enemy the target of the attack
     */
    fun attack(enemy: IUnit)
}

/**
 * This class represents a Unit from a generic game.
 *
 *
 * It's only used for teaching purposes
 *
 * @author Ignacio Slater Muñoz
 * @since 0.1.x01
 */
abstract class AbstractUnit(override var hitPoints: Int, override var strength: Int) : IUnit {

    protected fun attack(enemy: IUnit, damage: Int) {
        enemy.hitPoints -= damage
    }

}
